Set ROI origin to match well origin

tcompa opened the pull request:
- [x] Shift the FOV/well ROI tables so that their minimum XYZ values are at 0.
- [x] Remove workaround https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/339#issuecomment-1460020767
- [x] Adapt `zenodo_zarr` fixture to "fix" current zenodo datasets, and open issue about update (ref #526)
- [x] #460
- [x] #507
- [x] #525 
- [x] #527

Close #339
Close #460
Close #525
Close #507
Close #527


## Checklist before merging
- [x] I added an appropriate entry to `CHANGELOG.md`

tcompa commented:
This is mostly ready on my side, and it should streamline the understanding of ROIs from now on. FOV/well ROIs will have zero origin, and 

@jluethi a review would be useful.

I don't have a full understanding of what parts of registration tasks are tested, and then I am not sure of whether there remain open issues related to the ROI origins. Some notes:
1. I removed the `reset_origin` argument of `convert_ROI_table_to_indices`, and then I also removed `reset_origin=False` from `apply_registration_to_image.py` (see PR diff).
2. I did not modify `apply_registration_to_ROI_tables.py`, but I'm now wondering whether the use of `reset_origin(..)` in this task is still relevant or not.

jluethi commented:
Great, will review!

> I did not modify apply_registration_to_ROI_tables.py, but I'm now wondering whether the use of reset_origin(..) in this task is still relevant or not.

If all tables already come with a reset origin, then this is not necessary anymore. We only needed to run reset_origins because FOV & well ROIs had non-0 origins.

tcompa commented:
> If all tables already come with a reset origin, then this is not necessary anymore. We only needed to run reset_origins because FOV & well ROIs had non-0 origins.

This makes sense, thanks. I just removed it with https://github.com/fractal-analytics-platform/fractal-tasks-core/pull/524/commits/f15863d2bc6651a97fdccccaf304348cfa81695b.

`test_registration.py::test_multiplexing_registration` is passing, and uses `apply_registration_to_ROI_tables`. Is this enough as a check, or do we need new tests?

jluethi commented:
Hmm, good question. I think the tests would catch it if it calculates the registration wrong (because I have some checks on the amount of 0-padded pixels and such). But I'll run it on example 03 as part of the review and verify the results visually as well, just to be sure :)

jluethi commented on fractal_tasks_core/lib_regions_of_interest.py:
```diff
@@ -28,14 +28,17 @@
+        df:
+            Input dataframe, possibly prepared through
+            `parse_yokogawa_metadata`.
+        metadata:
```
What do we pass as metadata? I couldn't find an example of this being used. Was this the old idea to pass additional info to ROI tables like e.g. timepoint info that was in the Yokogawa metadata?

tcompa replied:
>  I couldn't find an example of this being used.

The default includes `time`. Is `time` a column which comes up in some yokogawa metadata parsing?

jluethi replied:
I think it's related to this: https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/25#issuecomment-1235544131

The time metadata can be relevant for some downstream analysis. We added it last summer to the parsing output from the yokogawa parser. Max brought up that he would need that.
I don't think it's ever been used so far. But yes, it would be to get that metadata from the Yokogawa parser into the obs of the ROI tables.

Let's leave this there for the time being

jluethi commented on fractal_tasks_core/lib_regions_of_interest.py:
```diff
@@ -453,20 +469,31 @@ def convert_indices_to_regions(
+        ROI_table: Original ROI table.
+        x_pos: Name of the column with X position of ROIs.
+        y_pos: Name of the column with Y position of ROIs.
+        z_pos: Name of the column with Z position of ROIs.
```
Returns:
    Anndata ROI table, with shifted-to-zero origin for positions

tcompa commented on fractal_tasks_core/lib_regions_of_interest.py:
```diff
@@ -209,19 +222,18 @@ def convert_ROI_table_to_indices(
+        level: Pyramid level.
+        coarsening_xy: Linear coarsening factor in the YX plane.
+        cols_xyz_pos: Column names for XYZ ROI positions.
+        cols_xyz_len: Column names for XYZ ROI edges.
```
Added:
```
    Returns:
        Nested list of indices. The main list has one item per ROI. Each ROI
        item is a list of six integers as in `[start_z, end_z, start_y, end_y,
        start_x, end_x]`. The array-index interval for a given ROI is
        `start_x:end_x` along X, and so on for Y and Z.
```

jluethi commented:
In local tests, I keep running into this error (both for example 01 and example 03):
```
Traceback (most recent call last):
  File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/yokogawa_to_ome_zarr.py", line 260, in <module>
    run_fractal_task(
  File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/_utils.py", line 79, in run_fractal_task
    metadata_update = task_function(**pars)
  File "pydantic/decorator.py", line 40, in pydantic.decorator.validate_arguments.validate.wrapper_function
  File "pydantic/decorator.py", line 134, in pydantic.decorator.ValidatedFunction.call
  File "pydantic/decorator.py", line 206, in pydantic.decorator.ValidatedFunction.execute
  File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/yokogawa_to_ome_zarr.py", line 173, in yokogawa_to_ome_zarr
    sample = imread(tmp_images.pop())
  File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/dask/array/image.py", line 49, in imread
    imread = imread or sk_imread
NameError: name 'sk_imread' is not defined
```
I do have Python 3.9.0 on that machine though. Given that our CI is fine, I'll have to look into updating python & retrying tomorrow

tcompa commented:
> In local tests, I keep running into this error (both for example 01 and example 03):
> 
> ```
> Traceback (most recent call last):
>   File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/yokogawa_to_ome_zarr.py", line 260, in <module>
>     run_fractal_task(
>   File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/_utils.py", line 79, in run_fractal_task
>     metadata_update = task_function(**pars)
>   File "pydantic/decorator.py", line 40, in pydantic.decorator.validate_arguments.validate.wrapper_function
>   File "pydantic/decorator.py", line 134, in pydantic.decorator.ValidatedFunction.call
>   File "pydantic/decorator.py", line 206, in pydantic.decorator.ValidatedFunction.execute
>   File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/fractal_tasks_core/tasks/yokogawa_to_ome_zarr.py", line 173, in yokogawa_to_ome_zarr
>     sample = imread(tmp_images.pop())
>   File "/Users/joel/Library/CloudStorage/Dropbox/Joel/BioVisionCenter/Code/fractal/fractal-demos/examples/server/FRACTAL_TASKS_DIR/.fractal/fractal_tasks_core0.11.0/venv/lib/python3.9/site-packages/dask/array/image.py", line 49, in imread
>     imread = imread or sk_imread
> NameError: name 'sk_imread' is not defined
> ```
> 
> I do have Python 3.9.0 on that machine though. Given that our CI is fine, I'll have to look into updating python & retrying tomorrow

The obvious comment is that you may be missing `scikit-image` in the fractal-tasks-core 0.11.0 venv, but why?
Just to make sure: did you include the `fractal-tasks` package extra? Because that's where `scikit-image` is defined.

Other than that, we can debug it further - but I don't see any connection with the current PR.

jluethi commented:
> Just to make sure: did you include the fractal-tasks package extra? Because that's where scikit-image is defined.

Ah, that was it! 🤦🏻‍♂️

Downside of trying to do testing on a rush. Interesting though to see that this is how the issue pops up if someone forgets to include the extras!

Retrying now :)

tcompa commented:
> The changes look good to me. Great that you updated docstrings @tcompa ! I added a few minor points to where we could clarify them further, but that's about it.

Thanks for the review. I fixed the minor points and I don't have anything more to add.

> I'll also run some tests on example 03, will report results :)

This would be the last check, and then I think we could merge this.

jluethi commented:
Ok, I can now confirm that the multiplexing examples still work as expected. PR approved :)
